File: tests/e2e/assert_partitioned_context.mjs
// Assert detailed partition identity and header isolation through the Node API.

import { readFile, writeFile } from "node:fs/promises";
import process from "node:process";

import * as rookieCookies from "../../bindings/node/index.js";
import { verifyCookieRecords } from "./cookie_manifest.mjs";

const [
  engine,
  database,
  browserId,
  topOrigin,
  otherTopOrigin,
  thirdOrigin,
  sourcePortArg,
  output,
] = process.argv.slice(2);

if (
  !["chromium", "firefox"].includes(engine) ||
  !database ||
  !topOrigin ||
  !otherTopOrigin ||
  !thirdOrigin ||
  !sourcePortArg
) {
  console.error(
    "usage: node assert_partitioned_context.mjs <chromium|firefox> <db> <browser-id-or-dash> <top-origin> <other-top-origin> <third-origin> <source-port> [output]",
  );
  process.exit(2);
}

const sourcePort = Number(sourcePortArg);
if (!Number.isSafeInteger(sourcePort) || sourcePort <= 0) {
  throw new Error(`invalid source port ${sourcePortArg}`);
}

function schemefulSite(origin) {
  const parsed = new URL(origin);
  const labels = parsed.hostname.split(".");
  if (labels.length !== 3 || !["top", "other"].includes(labels[0])) {
    throw new Error(`unexpected controlled top-level origin ${origin}`);
  }
  return `${parsed.protocol}//${labels.slice(1).join(".")}`;
}

const options = { path: database };
if (engine === "chromium") options.browserId = browserId;
const snapshot = await rookieCookies.fromPath(options);
const detailed = snapshot.detailedCookies.filter(({ cookie }) =>
  cookie.name.startsWith("rookie_"),
);

const byName = new Map();
for (const record of detailed) {
  const values = byName.get(record.cookie.name) || [];
  values.push(record);
  byName.set(record.cookie.name, values);
}
const expectedCounts = new Map([
  ["rookie_top", 2],
  ["rookie_chips", 3],
]);
if (engine === "firefox") expectedCounts.set("rookie_dfpi", 2);
if (
  byName.size !== expectedCounts.size ||
  [...expectedCounts].some(
    ([name, count]) => (byName.get(name) || []).length !== count,
  )
) {
  throw new Error(
    `context corpus mismatch: expected ${JSON.stringify(Object.fromEntries(expectedCounts))}, got ${JSON.stringify(Object.fromEntries([...byName].map(([name, records]) => [name, records.length])))}`,
  );
}
const rawManifestPath = process.env.ROOKIE_E2E_CONTEXT_MANIFEST;
const rawManifest = rawManifestPath
  ? JSON.parse(await readFile(rawManifestPath, "utf8"))
  : null;
if (rawManifestPath) {
  verifyCookieRecords(
    rawManifestPath,
    "detailed",
    detailed,
    `Node ${engine} raw context`,
  );
}

function exactlyTwo(name) {
  const records = byName.get(name) || [];
  if (records.length !== 2) {
    throw new Error(
      `expected exactly two colliding ${name} identities, got ${records.length}`,
    );
  }
  return records;
}

const top = exactlyTwo("rookie_top");
const chips = byName.get("rookie_chips");
if (engine === "chromium") {
  if (
    top.some(
      ({ context }) =>
        context.topFrameSiteKey != null && context.topFrameSiteKey !== "",
    )
  ) {
    throw new Error("first-party top cookie became partitioned");
  }
  const unpartitioned = chips.filter(
    ({ context }) =>
      context.topFrameSiteKey == null || context.topFrameSiteKey === "",
  );
  if (
    unpartitioned.length !== 1 ||
    unpartitioned[0].cookie.value !== "unpartitioned"
  ) {
    throw new Error(
      "Chromium lost the unpartitioned cookie sharing the CHIPS flat identity",
    );
  }
  const labels = new Set();
  for (const record of chips) {
    const key = record.context.topFrameSiteKey;
    if (key == null || key === "") continue;
    const label = ["a", "c"].find((candidate) =>
      String(key).includes(`rookie-${candidate}.test`),
    );
    if (!label || labels.has(label)) {
      throw new Error(`unexpected Chromium partition key ${key}`);
    }
    labels.add(label);
    if (record.cookie.value !== `partition-${label}`) {
      throw new Error(
        `Chromium partition ${label} carried ${record.cookie.value}`,
      );
    }
    if (record.context.hasCrossSiteAncestor !== true) {
      throw new Error("CHIPS row lost its cross-site ancestor bit");
    }
    if (record.context.sourcePort !== sourcePort) {
      throw new Error(`CHIPS source port was ${record.context.sourcePort}`);
    }
    if (record.context.sourceScheme !== 2) {
      throw new Error("CHIPS HTTPS source scheme was not 2");
    }
    if (record.context.isPersistent !== true) {
      throw new Error("CHIPS persistence bit was not true");
    }
  }
} else {
  const dfpi = exactlyTwo("rookie_dfpi");
  const unpartitioned = chips.filter(
    ({ context }) =>
      context.partitionKey == null || context.partitionKey === "",
  );
  if (
    unpartitioned.length !== 1 ||
    unpartitioned[0].cookie.value !== "unpartitioned"
  ) {
    throw new Error(
      "Firefox lost the unpartitioned cookie sharing the partitioned flat identity",
    );
  }
  for (const records of [chips, dfpi]) {
    const labels = new Set();
    for (const record of records) {
      if (
        record.cookie.name === "rookie_chips" &&
        (record.context.partitionKey == null ||
          record.context.partitionKey === "")
      ) {
        continue;
      }
      const label = ["a", "c"].find((candidate) =>
        String(record.context.partitionKey).includes(
          `rookie-${candidate}.test`,
        ),
      );
      if (!label || labels.has(label)) {
        throw new Error(
          `${record.cookie.name} has unexpected partition key ${record.context.partitionKey}`,
        );
      }
      labels.add(label);
      if (
        typeof record.context.originAttributes !== "string" ||
        !record.context.originAttributes.includes("partitionKey=")
      ) {
        throw new Error(
          `${record.cookie.name} lacks partitioned originAttributes`,
        );
      }
      if (![null, 0].includes(record.context.userContextId)) {
        throw new Error(
          `${record.cookie.name} unexpectedly entered a container`,
        );
      }
      if (![null, 0].includes(record.context.privateBrowsingId)) {
        throw new Error(
          `${record.cookie.name} unexpectedly entered private browsing`,
        );
      }
      const expected =
        record.cookie.name === "rookie_chips"
          ? `partition-${label}`
          : `dfpi-${label}`;
      if (record.cookie.value !== expected) {
        throw new Error(
          `${record.cookie.name} partition ${label} carried ${record.cookie.value}`,
        );
      }
    }
  }
}

const matchingContext = {
  url: `${thirdOrigin}/echo`,
  topLevelSite: schemefulSite(topOrigin),
  resource: "subresource",
  method: "safe",
};
const matching = snapshot.header(matchingContext);
const other = snapshot.header({
  ...matchingContext,
  topLevelSite: schemefulSite(otherTopOrigin),
});
const tokens = (header) =>
  header
    .split(";")
    .map((token) => token.trim())
    .filter(Boolean)
    .sort();
let expectedMatching = [
  "rookie_chips=partition-a",
  "rookie_chips=unpartitioned",
];
let expectedOther = ["rookie_chips=partition-c", "rookie_chips=unpartitioned"];
if (engine === "firefox") {
  expectedMatching.push("rookie_dfpi=dfpi-a");
  expectedOther.push("rookie_dfpi=dfpi-c");
}
if (rawManifest) {
  expectedMatching = rawManifest.expected_headers.matching;
  expectedOther = rawManifest.expected_headers.other_top_level_site;
}
if (
  JSON.stringify(tokens(matching)) !==
  JSON.stringify([...expectedMatching].sort())
) {
  throw new Error(
    `matching header set mismatch: expected ${JSON.stringify([...expectedMatching].sort())}, got ${JSON.stringify(tokens(matching))}`,
  );
}
if (
  JSON.stringify(tokens(other)) !== JSON.stringify([...expectedOther].sort())
) {
  throw new Error(
    `other header set mismatch: expected ${JSON.stringify([...expectedOther].sort())}, got ${JSON.stringify(tokens(other))}`,
  );
}

let missingSelector;
try {
  snapshot.header({
    url: `${thirdOrigin}/echo`,
    resource: "subresource",
    method: "safe",
  });
  throw new Error("partitioned snapshot accepted an incomplete context");
} catch (error) {
  if (error.rookieCode !== "incomplete_send_context") throw error;
  missingSelector = error.rookieCode;
}

const result = {
  engine,
  detailed: detailed.sort((left, right) =>
    JSON.stringify(left).localeCompare(JSON.stringify(right)),
  ),
  headers: {
    matching,
    otherTopLevelSite: other,
    missingSelector,
  },
};
const encoded = `${JSON.stringify(result, null, 2)}\n`;
if (output) {
  await writeFile(output, encoded, { encoding: "utf8", flag: "wx" });
} else {
  process.stdout.write(encoded);
}

File: tests/e2e/assert_partitioned_context.py
#!/usr/bin/env python3
"""Assert browser-produced partition context and send-time isolation."""

from __future__ import annotations

import argparse
import json
import os
from pathlib import Path
import sys
from typing import Any, Protocol

from cookie_manifest import ManifestError, load_manifest, verify_records


class ContextAssertionError(RuntimeError):
    """The detailed snapshot did not preserve browser isolation semantics."""


class Snapshot(Protocol):
    def detailed_cookies(self) -> list[dict[str, Any]]: ...

    def header(self, context: dict[str, Any]) -> str: ...


def normalized(records: list[dict[str, Any]]) -> list[dict[str, Any]]:
    return sorted(
        records,
        key=lambda record: (
            record["cookie"]["domain"],
            record["cookie"]["path"],
            record["cookie"]["name"],
            json.dumps(record.get("context", {}), sort_keys=True),
        ),
    )


def context_value(context: dict[str, Any], snake: str, camel: str) -> Any:
    return context.get(snake, context.get(camel))


def header_tokens(header: str) -> list[str]:
    return sorted(token.strip() for token in header.split(";") if token.strip())


def schemeful_site(origin: str) -> str:
    """Return the controlled test origin's browser-persisted schemeful site."""

    from urllib.parse import urlparse

    parsed = urlparse(origin)
    host = parsed.hostname or ""
    labels = host.split(".")
    if len(labels) != 3 or labels[0] not in {"top", "other"}:
        raise ContextAssertionError(f"unexpected controlled top-level origin {origin!r}")
    return f"{parsed.scheme}://{'.'.join(labels[1:])}"


def validate_context_snapshot(
    snapshot: Snapshot,
    *,
    engine: str,
    top_origin: str,
    other_top_origin: str,
    third_origin: str,
    expected_source_port: int,
    raw_manifest: Path | None = None,
) -> dict[str, Any]:
    detailed = [
        record
        for record in snapshot.detailed_cookies()
        if record.get("cookie", {}).get("name", "").startswith("rookie_")
    ]
    by_name: dict[str, list[dict[str, Any]]] = {}
    for record in detailed:
        by_name.setdefault(record["cookie"]["name"], []).append(record)

    expected_counts = {"rookie_top": 2, "rookie_chips": 3}
    if engine == "firefox":
        expected_counts["rookie_dfpi"] = 2
    actual_counts = {name: len(records) for name, records in by_name.items()}
    if actual_counts != expected_counts:
        raise ContextAssertionError(
            f"context corpus mismatch: expected {expected_counts}, got {actual_counts}"
        )
    manifest = load_manifest(raw_manifest) if raw_manifest is not None else None
    if manifest is not None:
        verify_records(
            manifest,
            "detailed",
            detailed,
            surface=f"Python {engine} raw context",
        )

    for required in ("rookie_top", "rookie_chips"):
        expected = expected_counts[required]
        if len(by_name.get(required, [])) != expected:
            raise ContextAssertionError(
                f"expected exactly {expected} colliding {required} identities, "
                f"got {len(by_name.get(required, []))}"
            )

    top_contexts = [record.get("context", {}) for record in by_name["rookie_top"]]
    if engine == "chromium":
        if any(
            context_value(context, "top_frame_site_key", "topFrameSiteKey")
            not in (None, "")
            for context in top_contexts
        ):
            raise ContextAssertionError("a first-party top cookie became partitioned")
        unpartitioned = [
            record
            for record in by_name["rookie_chips"]
            if context_value(
                record.get("context", {}), "top_frame_site_key", "topFrameSiteKey"
            )
            in (None, "")
        ]
        if (
            len(unpartitioned) != 1
            or unpartitioned[0]["cookie"]["value"] != "unpartitioned"
        ):
            raise ContextAssertionError(
                "Chromium lost the unpartitioned cookie sharing the CHIPS flat identity"
            )
        keyed: dict[str, dict[str, Any]] = {}
        for record in by_name["rookie_chips"]:
            chips_context = record.get("context", {})
            key = context_value(chips_context, "top_frame_site_key", "topFrameSiteKey")
            if key in (None, ""):
                continue
            label = next(
                (
                    candidate
                    for candidate in ("a", "c")
                    if f"rookie-{candidate}.test" in str(key)
                ),
                None,
            )
            if label is None or label in keyed:
                raise ContextAssertionError(
                    f"unexpected Chromium partition keys: {key!r}"
                )
            keyed[label] = record
            if (
                context_value(
                    chips_context, "has_cross_site_ancestor", "hasCrossSiteAncestor"
                )
                is not True
            ):
                raise ContextAssertionError(
                    "CHIPS row lost its cross-site ancestor bit"
                )
            if (
                context_value(chips_context, "source_port", "sourcePort")
                != expected_source_port
            ):
                raise ContextAssertionError(
                    f"CHIPS source port was {context_value(chips_context, 'source_port', 'sourcePort')!r}"
                )
            if context_value(chips_context, "source_scheme", "sourceScheme") != 2:
                raise ContextAssertionError("CHIPS HTTPS source scheme was not 2")
            if (
                context_value(chips_context, "is_persistent", "isPersistent")
                is not True
            ):
                raise ContextAssertionError("CHIPS persistence bit was not true")
        for label, record in keyed.items():
            if record["cookie"]["value"] != f"partition-{label}":
                raise ContextAssertionError(
                    f"Chromium partition {label} carried {record['cookie']['value']!r}"
                )
    else:
        if len(by_name.get("rookie_dfpi", [])) != 2:
            raise ContextAssertionError(
                f"expected two Firefox dFPI rows, got {len(by_name.get('rookie_dfpi', []))}"
            )
        chips_unpartitioned = [
            record
            for record in by_name["rookie_chips"]
            if context_value(record.get("context", {}), "partition_key", "partitionKey")
            in (None, "")
        ]
        if (
            len(chips_unpartitioned) != 1
            or chips_unpartitioned[0]["cookie"]["value"] != "unpartitioned"
        ):
            raise ContextAssertionError(
                "Firefox lost the unpartitioned cookie sharing the partitioned flat identity"
            )
        for name in ("rookie_chips", "rookie_dfpi"):
            labels: set[str] = set()
            for record in by_name[name]:
                context = record.get("context", {})
                origin_attributes = context_value(
                    context, "origin_attributes", "originAttributes"
                )
                partition_key = context_value(context, "partition_key", "partitionKey")
                if name == "rookie_chips" and partition_key in (None, ""):
                    continue
                label = next(
                    (
                        candidate
                        for candidate in ("a", "c")
                        if f"rookie-{candidate}.test" in str(partition_key)
                    ),
                    None,
                )
                if label is None or label in labels:
                    raise ContextAssertionError(
                        f"{name} has unexpected parsed partition key {partition_key!r}"
                    )
                labels.add(label)
                if (
                    not isinstance(origin_attributes, str)
                    or "partitionKey=" not in origin_attributes
                ):
                    raise ContextAssertionError(
                        f"{name} lacks complete partitioned originAttributes"
                    )
                if context_value(context, "user_context_id", "userContextId") not in (
                    None,
                    0,
                ):
                    raise ContextAssertionError(
                        f"{name} unexpectedly entered a container"
                    )
                if context_value(
                    context, "private_browsing_id", "privateBrowsingId"
                ) not in (None, 0):
                    raise ContextAssertionError(
                        f"{name} unexpectedly entered private browsing"
                    )
                expected_value = (
                    f"partition-{label}" if name == "rookie_chips" else f"dfpi-{label}"
                )
                if record["cookie"]["value"] != expected_value:
                    raise ContextAssertionError(
                        f"{name} partition {label} carried {record['cookie']['value']!r}"
                    )

    matching_context = {
        "url": f"{third_origin}/echo",
        "top_level_site": schemeful_site(top_origin),
        "resource": "subresource",
        "method": "safe",
    }
    other_context = {
        **matching_context,
        "top_level_site": schemeful_site(other_top_origin),
    }
    matching_header = snapshot.header(matching_context)
    other_header = snapshot.header(other_context)
    expected_matching = [
        "rookie_chips=partition-a",
        "rookie_chips=unpartitioned",
    ]
    expected_other = [
        "rookie_chips=partition-c",
        "rookie_chips=unpartitioned",
    ]
    if engine == "firefox":
        expected_matching.append("rookie_dfpi=dfpi-a")
        expected_other.append("rookie_dfpi=dfpi-c")
    if manifest is not None:
        expected_matching = manifest["expected_headers"]["matching"]
        expected_other = manifest["expected_headers"]["other_top_level_site"]
    if header_tokens(matching_header) != sorted(expected_matching):
        raise ContextAssertionError(
            f"matching header set mismatch: expected {sorted(expected_matching)!r}, "
            f"got {header_tokens(matching_header)!r}"
        )
    if header_tokens(other_header) != sorted(expected_other):
        raise ContextAssertionError(
            f"other header set mismatch: expected {sorted(expected_other)!r}, "
            f"got {header_tokens(other_header)!r}"
        )

    try:
        snapshot.header(
            {
                "url": f"{third_origin}/echo",
                "resource": "subresource",
                "method": "safe",
            }
        )
    except Exception as error:  # public bindings expose a typed extension error
        code = getattr(error, "code", None)
        if code != "incomplete_send_context" and "incomplete_send_context" not in str(
            error
        ):
            raise ContextAssertionError(
                f"missing selector failed with the wrong error: {error!r}"
            ) from error
    else:
        raise ContextAssertionError(
            "partitioned snapshot accepted an incomplete context"
        )

    return {
        "engine": engine,
        "detailed": normalized(detailed),
        "headers": {
            "matching": matching_header,
            "other_top_level_site": other_header,
            "missing_selector": "incomplete_send_context",
        },
    }


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--engine", choices=("chromium", "firefox"), required=True)
    parser.add_argument("--database", type=Path, required=True)
    parser.add_argument("--browser-id", default="chromium")
    parser.add_argument("--top-origin", required=True)
    parser.add_argument("--other-top-origin", required=True)
    parser.add_argument("--third-origin", required=True)
    parser.add_argument("--source-port", type=int, required=True)
    parser.add_argument("--output", type=Path)
    args = parser.parse_args()

    try:
        import rookie_cookies

        if args.engine == "chromium":
            snapshot = rookie_cookies.from_path(
                str(args.database), browser_id=args.browser_id
            )
        else:
            snapshot = rookie_cookies.from_path(str(args.database))
        result = validate_context_snapshot(
            snapshot,
            engine=args.engine,
            top_origin=args.top_origin,
            other_top_origin=args.other_top_origin,
            third_origin=args.third_origin,
            expected_source_port=args.source_port,
            raw_manifest=(
                Path(os.environ["ROOKIE_E2E_CONTEXT_MANIFEST"])
                if os.environ.get("ROOKIE_E2E_CONTEXT_MANIFEST")
                else None
            ),
        )
        encoded = (
            json.dumps(result, ensure_ascii=False, sort_keys=True, indent=2) + "\n"
        )
        if args.output is None:
            print(encoded, end="")
        else:
            args.output.write_text(encoded, encoding="utf-8")
    except (ContextAssertionError, ManifestError, OSError, ValueError) as error:
        print(f"partition context assertion failed: {error}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())
